blackjax/mcmc/proposal.py

```diff
@@ -65,26 +66,55 @@ def update(initial_energy: float, state: IntegratorState) -> Tuple[Proposal, boo
             The new state.
 
         """
-        new_energy = -state.logdensity + kinetic_energy(state.momentum)
-
-        delta_energy = initial_energy - new_energy
-        delta_energy = jnp.where(jnp.isnan(delta_energy), -jnp.inf, delta_energy)
-        is_transition_divergent = jnp.abs(delta_energy) > divergence_threshold
-
-        # The weight of the new proposal is equal to H0 - H(z_new)
-        weight = delta_energy
-
-        # Acceptance statistic min(e^{H0 - H(z_new)}, 1)
-        sum_log_p_accept = jnp.minimum(delta_energy, 0.0)
-
-        return (
-            Proposal(
-                state,
-                new_energy,
-                weight,
-                sum_log_p_accept,
-            ),
-            is_transition_divergent,
+        new_energy = energy(state)
+        return proposal_from_energy_diff(
+            initial_energy, new_energy, divergence_threshold, state
+        )
+
+    return new, update
+
+
+def proposal_from_energy_diff(initial_energy, new_energy, divergence_threshold, state):
+    delta_energy = initial_energy - new_energy
+    delta_energy = jnp.where(jnp.isnan(delta_energy), -jnp.inf, delta_energy)
+    is_transition_divergent = jnp.abs(delta_energy) > divergence_threshold
+
+    # The weight of the new proposal is equal to H0 - H(z_new)
+    weight = delta_energy
+
+    # Acceptance statistic min(e^{H0 - H(z_new)}, 1)
+    sum_log_p_accept = jnp.minimum(delta_energy, 0.0)
+
+    return (
+        Proposal(
+            state,
+            new_energy,
+            weight,
+            sum_log_p_accept,
+        ),
+        is_transition_divergent,
+    )
+
+
+def transition_aware_proposal_generator(
+    initial_energy_fn: Callable,
+    transition_energy_fn: Callable,
+    divergence_threshold: float,
+    proposal_factory=proposal_from_energy_diff,
+) -> tuple[Callable, Callable]:
+    """
+    A proposal generator that takes into account the transition from one state to the
+    other in order to calculate energies.
+    """
+
+    def new(state: TrajectoryState) -> Proposal:
+        return Proposal(state, initial_energy_fn(state), 0.0, -np.inf)
+
+    def update(prev_state: float, new_state: TrajectoryState) -> tuple[Proposal, bool]:
```

Comment:
`prev_state` is a `TrajectoryState` not a `float`.

I prefer the naming:
`prev_state` -> `inital_state`
`new_state` -> `state`